src/config/builder.rs

```diff
@@ -392,8 +424,77 @@ mod tests {
     /// should ideally be able to fix so that the original hash passes!
     fn version_hash_match() {
         assert_eq!(
-            "2bc405edda02d6a32f2a93332d5e73840670cce4731a78cf6f73e2f7df8e7229",
+            "84112522217c90260692863950365f9149f87b723ceea7930eec863653d697c8",
             ConfigBuilder::default().sha256_hash()
         );
     }
+
```

Comment:
Can we add a test here that actually just uses `serde_json` directly and creates an object with unsorted keys, and then runs it through the same logic we use to get our sorted maps, and ensures that they're actually sorted? Might also make sense to extract the "obj to value, value to string" lines into their own free function so the test can use the same codepath as well.

I mostly don't feel comfortable about basing the correctness of the logic on whether or not a feature flag is enabled on a dependency.

Reply:
I'm not sure I follow 100%. Are you looking for a test like:
```rust
// for some iterations
  let mut map = HashMap::new();
  map.insert("hello", "world");
  map.insert("foo", "foov");
  map.insert("bar", "barv");
  map.insert("baz", "bazv");
  map.insert("car", "carv");
  assert_eq!({SOME ORDERED COMPARISON}, serde_json::to_value(&map).unwrap()));
```
That does seem slightly more resilient in case `loki` `labels` changes. Outside that, in terms of a feature flag changing for `serde` or some underlying `serde` implementation change, wouldn't both tests (current and proposed) accurately fail?

Reply:
While you're not wrong that the current test could also expose such an innocuous change, the problem in my eyes is that it's not deterministic. We're already running it through multiple iterations to try and catch a different iteration ordering on the maps we hope to sort, but that's still not guaranteed to catch it.

We can be absolutely certain that the code works if we simply construct an `IndexMap`, for example, with out of order keys and make sure they get sorted. We control the iteration order, we start with keys that are known to be unsorted, and we assert that the code does in fact end up sorting them correctly.

Reply:
Aha, I see, great point. Will do.

Reply:
Added in [3fcb368](https://github.com/vectordotdev/vector/pull/13355/commits/3fcb3681a3978e1f451c4144e18f994f607f362c)